src/V3Fork.cpp

```diff
@@ -34,12 +46,351 @@
+    int m_forkDepth = 0;  // Number of asynchronous forks we are currently under
+    bool m_afterTimingControl = false;  // A timing control might've be executed in the current
+                                        // process
+
+    // METHODS
+
+    DynScopeFrame* frameOf(AstNode* nodep) {
+        auto frameIt = m_frames.find(nodep);
+        if (frameIt == m_frames.end()) return nullptr;
+        return frameIt->second;
+    }
+
+    const DynScopeFrame* frameOf(AstNode* nodep) const {
+        auto frameIt = m_frames.find(nodep);
+        if (frameIt == m_frames.end()) return nullptr;
+        return frameIt->second;
+    }
+
+    DynScopeFrame* pushDynScopeFrame() {
+        DynScopeFrame* const frame = new DynScopeFrame{m_modp, m_procp};
+        auto r = m_frames.emplace(std::make_pair(m_procp, frame));
+        UASSERT_OBJ(r.second, m_modp, "Procedure already contains a frame");
+        return frame;
+    }
+
+    static string generateDynScopeClassName(AstNode* fromp) {
+        // TODO: Ensure no collisions occur
+        string n = "__VDynScope__" + (!fromp->name().empty() ? (fromp->name() + "__") : "ANON__")
+                   + cvtToHex(fromp);
+        return n;
+    }
+
+    static string generateDynScopeHandleName(AstNode* fromp) {
+        return "__VDynScope_" + (fromp->name().empty() ? cvtToHex(fromp) : fromp->name());
+    }
+
+    void replaceWithMemberSel(AstVarRef* refp, const DynScopeInstance& dynScope) {
+        VNRelinker handle;
+        refp->unlinkFrBack(&handle);
+        AstMemberSel* const membersel = new AstMemberSel{
+            refp->fileline(), new AstVarRef{refp->fileline(), dynScope.handlep, refp->access()},
+            refp->dtypep()};
+        membersel->name(refp->varp()->name());
+        if (VL_UNLIKELY(refp->varp()->direction() == VDirection::INPUT))
+            membersel->varp(
+                VN_AS(m_memberMap.findMember(dynScope.classp, refp->varp()->name()), Var));
+        else
+            membersel->varp(refp->varp());
+        handle.relink(membersel);
+        VL_DO_DANGLING(refp->deleteTree(), refp);
+    }
+
+    static bool hasAsyncFork(AstNode* nodep) {
+        bool afork = false;
+        nodep->foreach([&](AstFork* forkp) {
+            if (!forkp->joinType().join()) { afork = true; }
+        });
+        return afork;
+    }
+
+    void bindNodeToDynScope(AstNode* nodep, DynScopeFrame* frame) {
+        m_frames.emplace(std::make_pair(nodep, frame));
+    }
+
+    bool needsDynScope(const AstVarRef* refp) const {
+        // Conditions
+        auto mutated = [=]() -> bool {
+            return refp->varp()->isClassHandleValue() ? refp->user2()
+                                                      : refp->access().isWriteOrRW();
+        };
+        auto canEscapeTheScope = [=]() -> bool {
+            return (m_forkDepth > refp->varp()->user1()) && refp->varp()->isFuncLocal();
+        };
+        auto afterDelay = [=]() -> bool { return m_afterTimingControl; };
```

Comment:
Think these lambdas aren't required, can just write "normal" equation.

Reply:
I only split the equation into three separate steps to make it easier to understand. By using variables I gave them names that describe what those parts are responsible for. Then I wrapped them in lambdas to make short-circuiting still possible, despite that split.

Here's an alternative proposal, without any lambdas:

```cpp
        return
            // Can this variable escape the scope
            ((m_forkDepth > refp->varp()->user1()) && refp->varp()->isFuncLocal())
            && (
                // Is it mutated
                (refp->varp()->isClassHandleValue() ? refp->user2() : refp->access().isWriteOrRW())
                // Or is it after a timing-control event
                || m_afterTimingControl);
```

Let me know which one do you like more.

Reply:
Like the alternative, or just make a few "const bool" with separate intermediate terms if you want to be more expressive.

Reply:
Ok. I changed it to my alternative proposal.